crates/engine/src/ext.rs

```diff
@@ -0,0 +1,335 @@
+// Copyright 2018-2021 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! TODO
+//!
+//! Refer to substrate SRML contract module for more documentation.
+
+use super::{
+    exec_context::ExecContext,
+    test_api::EmittedEvent,
+};
+
+use core::cell::RefCell;
+use ink_primitives::Key;
+use std::collections::HashMap;
+
+type Result = core::result::Result<(), Error>;
+
+pub struct EnvInstance {
+    /// The environment storage.
+    pub storage: HashMap<Vec<u8>, Vec<u8>>,
+    /// Emitted events recorder.
+    pub emitted_events: Vec<EmittedEvent>,
+    /// Current execution context and context.
+    pub exec_context: Option<ExecContext>,
+    /// The caller.
+    pub caller: Vec<u8>,
```

Comment:
I am sure our `AccountId` in SEAL are always 32 bytes, so from that standpoint you should be fine restricting the off-chain environment to 32 bytes `AccountId` as well.

Reply:
Nvm, `Vec<u8>` is okay here.